feat(control-interface)!: rename max-concurrent to max-instances, simplify scale

brooksmtownsend opened the pull request:
## Feature or Problem
After a month or so of working with the changes from #696, I've come to a few conclusions about the developer experience of the new scaling model. All the thoughts below are up for discussion I just wanted to propose the changes.

1. This number `--max-concurrent` is describing "the maximum number of concurrent instances allowed to execute" and and I slightly prefer `--max-instances` to describe this number. This is also a more intuitive number when describing capability provider instances, and the eventual WASI provider instances, and will be better to replace the `replicas` field in wadm manifests. This might just be `instances`, but that can be determined later.
2. The difference between scaling to `None`, `Some(0)`, and `Some(10000)` is still confusing, and "unbounded" is a scary number.
3. It's a little weird to work with the `Option<NonZeroUsize>` type all over the place.

In order to address these issues, I made the following changes
1. Renamed from `--max-concurrent` / `max_concurrent` to `--max-instances` / `max_instances` around the codebase.
2. The maximum instances parameter is no longer an optional number, just a u32. In developer facing clients we'll default to `u16::MAX` and it will be easy to see that number.
    a. ~If we run into an issue in the future with an actor needing to scale past 65535 concurrent running instances you are welcome to @ me and I'll update everything to be u32s 😉~ It's u32 now 😉 
3. All max scale numbers are now `NonZeroUsize`s. I considered refactoring this to a `NonZeroU32` as well, but felt that was an unnecessary internal detail to correct before discussion.

## Related Issues
Addresses removing the singular events and adding `actor_scaled` re: #1092 

## Release Information
No rush on releasing these so they can go out with our next regular release.

control-interface v0.33.0
wash-lib v0.16.0
wash-cli v0.26.0
wasmCloud v0.82.0

## Consumer Impact
- control-interface has a few `Option<u16>` to `u32` changes, which should be simple and anyone previously passing `None` can simply pass `u32::MAX` to preserve the same behavior
- wash-lib has a few `Option<u16>` to `u32` changes, same as control-interface
- wash-cli won't have any breaking changes, only a name change in the flag (which has a backwards compatible alias)
- wasmCloud users will need to update their `wash-cli` as omitting the count will no longer scale to unbounded, it will stop the actor.

## Testing
<!---
Declare the testing information for this pull request
--->

### Unit Test(s)
Modified tests to use `u32`s instead of `Option<u16>`s 

### Acceptance or Integration
<!---
Indicate any changes or additions to the acceptance or integration test suite 
--->

### Manual Verification
<!---
Mandatory. Indicate the steps that you took to verify that this pull request works 
--->

connorsmith256 commented on Cargo.toml:
```diff
@@ -167,15 +167,15 @@ warp = { version = "0.3", default-features = false }
 wascap = { version = "0.12", path = "./crates/wascap", default-features = false }
 wash-cli = { version = "0", path = "./crates/wash-cli", default-features = false }
-wash-lib = { version = "0.15", path = "./crates/wash-lib", default-features = false }
+wash-lib = { version = "0.16", path = "./crates/wash-lib", default-features = false }
```
Are we ready to release these dependent crates? I ask because e.g. if we needed to do a patch bump of wash-cli, and we've changed the versions here but haven't published them, then publishing wash-cli would fail, right?

brooksmtownsend replied:
Very true. I don't think we need to release them yet, but I do think that we need to update the versions in these crates because there's new functionality. Put another way, if the newly updated code for `wash-cli` tries to compile against `wash-lib 0.15`, it will fail.

With this method of development in the monorepo, I think we have to bump crates when making breaking changes or pay attention to the conventional commits leading up to a release. This change going in means wash-cli has to bump a minor version, so we can't publish a patch anyways unless we cherry pick this commit out. Does that make sense? I want to make sure I'm thinking about this the right way since we should all take the same actions when working on changes that affect multiple libraries in this repo.

connorsmith256 commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -70,6 +58,17 @@ use wasmcloud_runtime::capability::{
 use wasmcloud_runtime::Runtime;
 use wasmcloud_tracing::context::TraceContextInjector;
 
+use crate::{
```
Why the reorder here?

brooksmtownsend replied:
It was automatic, maybe new rust-analyzer? I can move back but this seems closer in line with the convention we use, being
1. core
2. std
3. external
4. internal

connorsmith256 replied:
The order I saw elsewhere was
1. internal (`crate`)
2. core
3. std
4. external

FWIW I prefer the order you have now. I wish we had consistent tooling to do this automatically so I could stop thinking about it 😅

vados-cosmonic replied:
Just a data point but I personally prefer the order that brooks mentioned:

core 
std
external
Internal

connorsmith256 commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -3928,7 +3917,7 @@ impl Host {
     }
 
-    #[instrument(level = "trace", skip_all, fields(subject = %message.subject))]
+    #[instrument(level = "trace", skip_all, fields(subject = % message.subject))]
```
oops?

connorsmith256 commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -4308,7 +4297,7 @@ impl Host {
     }
 
-    #[instrument(level = "trace", skip_all, fields(bucket = %entry.bucket, key = %entry.key, revision = %entry.revision, operation = ?entry.operation))]
+    #[instrument(level = "trace", skip_all, fields(bucket = % entry.bucket, key = % entry.key, revision = % entry.revision, operation = ? entry.operation))]
```
oops?

connorsmith256 reviewed (commented):
LGTM, but a few questions about possibly-unintended changes

Also, for consideration: 

> If we run into an issue in the future with an actor needing to scale past 65535 concurrent running instances you are welcome to @ me and I'll update everything to be u32s 😉

wanna just make everything a u32 now? 🙂

brooksmtownsend commented on crates/wash-lib/Cargo.toml:
```diff
@@ -29,10 +29,10 @@ cli = [
 
 [package.metadata.docs.rs]
-no-default-features = true # skip building adapters (which requires internet access)
+no-default-features = true             # skip building adapters (which requires internet access)
```
I didn't do this manually it was just my editor auto format, unfortunately I'm having trouble finding the setting that caused this, will try and correct. Definitely don't want inconsistent styling

brooksmtownsend commented:
Just a note that I feel I've addressed the comments here, and I found out why the test is failing 😬 For backwards compatibility with previous events we publish an `actor_started` event for each actor scaled up. Now, by default, the behavior is scaling to `u32::MAX` with wash and the test, and so 4,294,967,296 `actor_started` events get published.

I think I'll wait to finish this one up until #1092 is done, which would be a nice change and resolve this issue. Alternatively I could go ahead and remove the `actor_started` event here, but that's half the work.

brooksmtownsend commented:
Added another commit which adds the `actor_scaled` event and removes the singular actor events. This was necessary as the default behavior in this PR with wash is to scale to `u32::MAX` actors, which would publish `u32::MAX` `actor_started` events 😅

vados-cosmonic commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -2380,7 +2381,7 @@ impl Host {
         annotations: &Annotations,
         actor_ref: impl AsRef<str>,
-        max: Option<NonZeroUsize>,
+        max: NonZeroUsize,
```
I guess there's no use case for forcing actor instantiation but having no max? I might have thought that would be a case for some sort of "preloading"

Also, did we miss `max_instances` rename here?

brooksmtownsend replied:
Yeah this was just shorthand, I can go refactor for clarity, no need to have ambiguity.

What I was thinking is that if we forced instantiation with no max, it would only ever be different if there were more than 2^32 concurrently executing instances of a single component. The benefit of getting rid of the option is clarity, no need to speculate between the difference of `None` and `Some(0)`

vados-cosmonic replied:
ah OK -- yeah that's definitely a simplification, and I guess if we ever want to have some sort of preload operation we can add one explicitly

vados-cosmonic reviewed (approved):
LGTM :rocket:

connorsmith256 commented on crates/host/src/wasmbus/event.rs:
```diff
@@ -36,87 +36,91 @@ fn format_actor_claims(claims: &jwt::Claims<jwt::Actor>) -> serde_json::Value {
 ) -> serde_json::Value {
     json!({
         "public_key": claims.subject,
-        "instance_id": instance_id,
```
Are any clients depending on this?

brooksmtownsend replied:
The diff here is actually really misleading, I only removed `instance_id` from actor_started and actor_stopped which no longer exist.

connorsmith256 commented on crates/wash-lib/Cargo.toml:
```diff
@@ -29,7 +29,7 @@ cli = [
 ]
 nats = ["async-nats", "wadm"]
-build-adapters = [ "wasmcloud-component-adapters/build-adapters" ]
+build-adapters = ["wasmcloud-component-adapters/build-adapters"]
```
I'd try to avoid changing this line since https://github.com/wasmCloud/wasmCloud/pull/1146 touches it too

brooksmtownsend replied:
Gah, I don't even have a formatter installed anymore, how does this happen

connorsmith256 commented on crates/wash-cli/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "wash-cli"
-version = "0.25.0"
+version = "0.24.1"
```
whoops

connorsmith256 commented on crates/wash-lib/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "wash-lib"
-version = "0.16.0"
+version = "0.15.0"
```
whoops

connorsmith256 commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -1597,7 +1598,8 @@ pub struct Host {
     links: RwLock<HashMap<String, LinkDefinition>>,
-    actor_claims: Arc<RwLock<HashMap<String, jwt::Claims<jwt::Actor>>>>, // TODO: use a single map once Claims is an enum
+    actor_claims: Arc<RwLock<HashMap<String, jwt::Claims<jwt::Actor>>>>,
+    // TODO: use a single map once Claims is an enum
```
Odd that this got shoved between the two lines

connorsmith256 commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -4264,78 +4260,86 @@ impl Host {
 
-    /// Publishes `count` `actor_started` events and one `actors_started` event with the supplied count
-    async fn publish_actor_started_events(
+    // TODO(#1092): only publish actor_scale_failed event after wasmCloud releases 0.81
```
Now that 0.81 is out, can we resolve these TODOs?

brooksmtownsend replied:
I actually have to change these to 0.82 😢 The idea was to get these events published but not remove the old ones until a minor bump is done

connorsmith256 reviewed (changes requested):
Just need to fix a couple version decrements
